EZP-28718: Embedded images should be handled as images instead of plain embed object after migrating from XmlText to RichText

vidarl opened the pull request:
This one fixes 
[EZP-28718](https://jira.ez.no/browse/EZP-28718) : Embedded images should be handled as images instead of plain embed object after migrating from XmlText to RichText
[EZP-28767](https://jira.ez.no/browse/EZP-28767) : Create a script to fix embedded images broken after migrating from XmlText to RichText

Running `bin/console ezxmltext:convert-to-richtext`  will ensure embedded images are tagged correctly in the resulting ezrichtext.

Running `bin/console ezxmltext:convert-to-richtext --fix-embedded-images-only` will only fix the embedded images on an already converted DB.

Optionally, use `--image-content-types=image,thumbnail,foobar` to specify which classes should considered as image classes. 


Related doc pr :
https://github.com/ezsystems/developer-documentation/pull/268

andrerom commented on bundle/Command/ConvertXmlTextToRichTextCommand.php:
```diff
@@ -64,25 +70,108 @@ protected function configure()
+    protected function loginAsAdmin()
+    {
+        $userService = $this->getContainer()->get('ezpublish.api.service.user');
+        $permissionResolver = $this->getContainer()->get('date_based_publisher.permission_resolver');
```
date_based_publisher?

vidarl replied:
Good catch. cut&paste error.. Fixed

andrerom commented on bundle/Command/ConvertXmlTextToRichTextCommand.php:
```diff
@@ -64,25 +70,108 @@ protected function configure()
+                'image-content-types',
+                null,
+                InputOption::VALUE_OPTIONAL,
+                'Comma separated list of content types which are considered as images when converting embedded tags. Default value is 27'
```
would be better if we use identifiers, type id's have a tendency to have different values depending on install more often then identifiers.

However the command can translate it to id using dbal, so service can stay on using type id.

vidarl replied:
ok. fixed

andrerom commented on bundle/Command/ConvertXmlTextToRichTextCommand.php:
```diff
@@ -120,6 +126,31 @@ protected function execute(InputInterface $input, OutputInterface $output)
+                    'identifier',
+                    $literalIdentifiers
+                )
+            );
```
You can simplify this, welcome to DBAL's PARAM_INT_ARRAY bind type.

ref: https://github.com/ezsystems/ezpublish-kernel/blob/d1143080d2325cadc6ccc1e3aee55b29aecebae4/eZ/Publish/Core/FieldType/BinaryBase/BinaryBaseStorage/Gateway/DoctrineStorage.php#L269

vidarl replied:
aha.. thanx.. Cause my solution was indeed insanely cumbersome

andrerom commented on bundle/Command/ConvertXmlTextToRichTextCommand.php:
```diff
@@ -120,6 +126,31 @@ protected function execute(InputInterface $input, OutputInterface $output)
+
+        $statement = $query->execute();
+        $result = array();
+        while ($row = $statement->fetch(PDO::FETCH_ASSOC)) {
```
afaik you can use `return $statement->fetchAll(PDO::FETCH_COLUMN);` to get what you want here.

andrerom commented on bundle/Command/ConvertXmlTextToRichTextCommand.php:
```diff
@@ -120,6 +126,31 @@ protected function execute(InputInterface $input, OutputInterface $output)
+        }
+
+        $query->select('id')
+            ->from($this->dbal->quoteIdentifier('ezcontentclass'))
```
Regarding this, I was about to say you don't need to quote constants with DBAL, but seeing it is done in https://github.com/ezsystems/ezpublish-kernel/blob/d1143080d2325cadc6ccc1e3aee55b29aecebae4/eZ/Publish/Core/FieldType/BinaryBase/BinaryBaseStorage/Gateway/DoctrineStorage.php#L269 I'm a bit unsure.


@alongosz Did you check that when writing it? Or was it left over from style of zeta based older db code?

vidarl replied:
You are right Andre, those constants does not need to be quoted it seems

andrerom replied:
> But it's up to you, for sure it would be more readable w/o quoting :)

It's not just up to me, but I prefer readability as long as we are whit-in:
- https://www.doctrine-project.org/projects/doctrine-dbal/en/2.7/reference/portability.html#portability
- https://www.doctrine-project.org/projects/doctrine-dbal/en/2.7/reference/security.html#security

andrerom replied:
So proposal on tables and columns: Write like this so we never need to be in dubt, and basically never need to quote:
```
    $query->select('c.id')
    ->from('ezcontentclass', 'c')
```

Avoids the need to do `$this->dbal->quoteIdentifier('id')` which is done further down here, also makes it explicit which table we act on all over. Again, optimizing for reading, cause that is what we do most of the time ;)

vidarl commented:
Fixed all review comments I believe...

vidarl commented:
Added more bugfixes:
- added support for `<embed node_id=...>` (previously, only `<embed object_id=...>` was supported
- Gracefully handling of provided node_id or object_id do not exists

andrerom commented on lib/FieldType/XmlText/Converter/RichText.php:
```diff
@@ -24,14 +25,29 @@ class RichText implements Converter
+     */
+    private $currentContentFieldId;
+
+    public function __construct(LoggerInterface $logger = null, $apiRepository = null, $imageContentTypes = array())
```
apiRepository should be first as it's not really optional.

imageContentTypes should be either optional setter injection, or required constructor argument. Rather then both at the same time.

andrerom commented on lib/FieldType/XmlText/Converter/RichText.php:
```diff
@@ -76,31 +100,104 @@ protected function reportNonUniqueIds(DOMDocument $document, $contentObjectAttri
+                try {
+                    $location = $locationService->loadLocation($id);
+                } catch (NotFoundException $e) {
+                    $this->logger->warning("Unable to find node_id=$id, referred to in embedded tag in contentobject_attribute.id=$this->currentContentFieldId.");
```
this and usage below should check for null on `$this->logger`, or code should be changed to set a NullLogger as default option if none is provided to avoid the logic.

andrerom commented on lib/FieldType/XmlText/Converter/RichText.php:
```diff
@@ -76,31 +100,104 @@ protected function reportNonUniqueIds(DOMDocument $document, $contentObjectAttri
+        $nodes = $xpath->query('//doc:ezembed');
+        foreach ($nodes as $node) {
+            //href is in format : ezcontent://123 or ezlocation://123
+            $href=$node->attributes->getNamedItem('href')->nodeValue;
```
` = `

andrerom commented on lib/FieldType/XmlText/Converter/RichText.php:
```diff
@@ -76,31 +100,104 @@ protected function reportNonUniqueIds(DOMDocument $document, $contentObjectAttri
+            //href is in format : ezcontent://123 or ezlocation://123
+            $href=$node->attributes->getNamedItem('href')->nodeValue;
+            $isContentId = strpos($href, 'ezcontent') === 0;
+            $id = (int) substr($href, strrpos($href, '/')+1);
```
We should setup cs checking on travis here, but it won't run on this PR right now as it's against dbal branch :)

vidarl replied:
Yeah.. CS and running unit tests would be good. Who can setup that? I guess it should be adressed in a separate PR though ....

andrerom commented on lib/FieldType/XmlText/Converter/RichText.php:
```diff
@@ -76,31 +100,104 @@ protected function reportNonUniqueIds(DOMDocument $document, $contentObjectAttri
+            } else {
+                if (($classAttribute !== null) && ($node->attributes->getNamedItem('class')->nodeValue === 'ez-embed-type-image')) {
+                    $node->removeAttribute('ezxhtml:class');
+                    //$node->setAttribute('ezxhtml:class', 'ez-embed-type-image');
```
to be removed?

As for line above, also the class logic here forget it can be several classes like in html. Or? (OE probably did not allow it, but I1'm thinking maybe the underlying xmltext allowed it if edited raw).
At least assume it can be several when comparing richtext nodes, as it adds `ez-embed-type-image` to whatever was there from before.

andrerom replied:
yes and yes :)

vidarl commented:
Fixed according to all review comments, and rebased

vidarl commented on lib/FieldType/XmlText/Converter/RichText.php:
```diff
@@ -76,31 +106,167 @@ protected function reportNonUniqueIds(DOMDocument $document, $contentObjectAttri
+        $xpath = new DOMXPath($richtextDocument);
+        $ns = $richtextDocument->documentElement->namespaceURI;
+        $xpath->registerNamespace('doc', $ns);
+        $nodes = $xpath->query('//doc:ezembed');
```
I have to deal with ezembed-inline too....

vidarl replied:
fixed

andrerom commented on bundle/Command/ConvertXmlTextToRichTextCommand.php:
```diff
@@ -165,18 +285,58 @@ protected function convertFields($dryRun, $contentObjectId, $checkDuplicateIds,
+                    ':version'
+                )
+            )
+            ->setParameters(array(
```
`[`

andrerom commented on lib/FieldType/XmlText/Converter/RichText.php:
```diff
@@ -24,14 +27,33 @@ class RichText implements Converter
+        if ($logger === null) {
+            $this->logger = new NullLogger();
+        } else {
+            $this->logger = $logger;
```
Could also be:
```
$this->logger = $logger instanceof LoggerInterface ? $logger : new NullLogger(); 
```

vidarl replied:
fixed

andrerom commented on bundle/Command/ConvertXmlTextToRichTextCommand.php:
```diff
@@ -27,12 +28,18 @@ class ConvertXmlTextToRichTextCommand extends ContainerAwareCommand
         parent::__construct();
 
         $this->dbal = $dbal;
         $this->logger = $logger;
```
Seems we need NullLogger here to form the looks of the new code bellow.

vidarl replied:
I'll instead remove default value from constructor, forcing a logger to be set in DI config
ok?

andrerom replied:
I think that is ok, but maybe double check with @adamwojs if there are still cases logger might not be set, as we tend to some places still allow it to be null for some reason. But that might just be baggage from earlier symfony versions, or attempt to have it as optional dependency.

andrerom commented on lib/FieldType/XmlText/Converter/RichText.php:
```diff
@@ -76,31 +106,167 @@ protected function reportNonUniqueIds(DOMDocument $document, $contentObjectAttri
+     */
+    public function convert(DOMDocument $inputDocument, $checkDuplicateIds = false, $contentFieldId = null)
     {
+        $this->currentContentFieldId = $contentFieldId;
```
could remove this as class property and rather inject it, to avoid global state set by method which is not a setter.

vidarl replied:
ok, passing on $contentFieldId down the call stack

andrerom commented on tests/lib/FieldType/Converter/RichTextTest.php:
```diff
@@ -56,6 +61,25 @@ public function providerForTestConvert()
+     */
+    public function providerForTestTagEmbeddedImages()
+    {
+        $map = array();
```
`[]`

vidarl replied:
fixed

andrerom commented on tests/lib/FieldType/Converter/RichTextTest.php:
```diff
@@ -92,4 +150,31 @@ public function testConvert($inputFilePath, $outputFilePath)
+
+        $inputDocument = $this->createDocument($inputFilePath);
+        $richText = new RichText($apiRepositoryStub);
+        $richText->setImageContentTypes(array(27));
```
`[]`

vidarl replied:
fixed

andrerom replied:
it's not it seems :)

vidarl commented:
review ping. PR should now be in compliance with all review comments

andrerom commented:
@alongosz Not sure what you are talking about here, this is migration scenario, it's something people run before booting the system normally, during upgrade. aAd If they don't they need to clear all cache anyway.

So while it would be great if we had generic migrations features supported somehow cleanly via SPI _(Storage Engine and FieldTypes at least)_, it's far far beyond the scope here and more for you and engineering  team to think about for future versions _(this is for existing, real world versions ;))_.

vidarl commented:
In general, there could be benefits with using persistence handler in  upgrade scripts. But there are pitfalls too. Often, upgrade scripts fix an invalid DB state ( schema change for instance ). So, having API calls that are able to deal with such things would be nice to have, but not sure it is worth the effort.

my 2 cents...

alongosz commented on tests/lib/FieldType/Converter/_fixtures/tag_embedded_images/input/002-add-ez-embed-type-image.xml:
```diff
@@ -0,0 +1,31 @@
+            <ezvalue key="limit">5</ezvalue>
+        </ezconfig>
+    </ezembed>
+</section>
```
Missing Line Feed at EOF

alongosz commented on tests/lib/FieldType/Converter/_fixtures/tag_embedded_images/input/003-remove-ez-embed-type-image.xml:
```diff
@@ -0,0 +1,31 @@
+            <ezvalue key="limit">5</ezvalue>
+        </ezconfig>
+    </ezembed>
+</section>
```
Same as above

alongosz commented on tests/lib/FieldType/Converter/_fixtures/tag_embedded_images/output/001-no-change.xml:
```diff
@@ -0,0 +1,23 @@
+            <ezvalue key="limit">5</ezvalue>
+        </ezconfig>
+    </ezembed>
+</section>
```
And here

alongosz commented on tests/lib/FieldType/Converter/_fixtures/tag_embedded_images/output/002-add-ez-embed-type-image.xml:
```diff
@@ -0,0 +1,31 @@
+            <ezvalue key="limit">5</ezvalue>
+        </ezconfig>
+    </ezembed>
+</section>
```
Here

alongosz commented on tests/lib/FieldType/Converter/_fixtures/tag_embedded_images/output/003-remove-ez-embed-type-image.xml:
```diff
@@ -0,0 +1,31 @@
+            <ezvalue key="limit">5</ezvalue>
+        </ezconfig>
+    </ezembed>
+</section>
```
and finally here :)

vidarl replied:
Missing LF at EOF fixed
